components/dialogs/dialogs.js

```diff
@@ -37,15 +38,24 @@ components.register('TooManyFilesDialogViewModel', function(args) {
 });
 
 function DialogViewModel(title) {
-  this.closed = new signals.Signal();
+  var self = this;
+  this.onclose = null;
   this.title = ko.observable(title);
   this.taDialogName = ko.observable('');
+  this.closePromise = new Promise(function(resolve) {
+    self.onclose = resolve;
+  });
+}
+DialogViewModel.prototype.closeThen = function(thenFunc) {
+  this.closePromise = this.closePromise.then(thenFunc);
+  return this;
 }
 DialogViewModel.prototype.setCloser = function(closer) {
   this.close = closer;
 }
-DialogViewModel.prototype.onclose = function() {
-  this.closed.dispatch();
+DialogViewModel.prototype.publish = function() {
```

Comment:
What about `show` instead of `publish`?

You could also return the `closePromise` instead of `this`. So it would be used like this:

```js
diag.show().then(function() { /* closed */ });
```

And `closeThen` can be removed.